### PR TITLE
WIP - #8 - Feature/8

### PR DESCRIPTION
**Pull request by OksanaSam**
#8 
## Description
Create the first five filters on the search page, as per specs and designs (https://xd.adobe.com/view/3b2fe38c-6d70-4a36-adab-e18285a668f0-6e7c/screen/9ec9e6d1-8442-49d2-98fd-44a0fbfe6137?fullscreen)

## Acceptance Criteria
User should be able to make requests to backend API
User should be able to adjust the request based on the filters needed.

## Added Dependencies
MSW - https://mswjs.io/

TO DO:
- Accordion to be made controlled
- Refactor local state + filtering

## UI
Screen Shot 2021-01-19 at 7 52 53 PM

## Notes
- msw mock fetch throws an error on the first render, then it's working fine
- MUI styles (color + width) on accordions  break on refresh

### CODE REVIEW

File: web-client/src/pages/drivers-search.tsx [hunk 1]
```diff
@@ -9,7 +9,7 @@ const Home: NextPage = () => {
   return (
     <>
       <Head>
-        <title>Search for volunteer drivers</title>
+        <title>Hello drivers</title>
```
Comment (l2silver) [inline web-client/src/pages/drivers-search.tsx:12]: Why did you change the title?

File: web-client/src/shared/components/drivers-search/Filters/filterData.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
+export const languageFilterdata = [
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/Filters/filterData.js:1]: So we should change this to a .ts file , so that it can be properly typechecked. And I think I would change the name from filterData.js to defaultValues.ts , because when I looked over the file, they looked more like defaultValues than anything else.

Comment (l2silver) [reply web-client/src/shared/components/drivers-search/Filters/filterData.js:1]: But I really like how you organized this. Very nice.

File: web-client/src/shared/components/drivers-search/Filters/index.tsx [hunk 3]
```diff
@@ -1,7 +1,284 @@
...
+
+  // const { loading, hasErrors, drivers } = useSelector<RootState>((state: RootState) => {
+  //     return {
+  //         loading: state.loading,
+  //         hasErrors: state.hasErrors,
+  //         drivers: state.drivers
+  //     }
+  // });
+
+  const { loading, hasErrors, drivers } = useSelector(selectDrivers);
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/Filters/index.tsx:53]: perfect!

File: web-client/src/shared/components/drivers-search/Filters/index.tsx [hunk 4]
```diff
@@ -1,7 +1,284 @@
...
+  //     drivers: Driver[];
+  // }
+
+  // const { loading, hasErrors, drivers } = useSelector<RootState>((state: RootState) => {
+  //     return {
+  //         loading: state.loading,
+  //         hasErrors: state.hasErrors,
+  //         drivers: state.drivers
+  //     }
+  // });
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/Filters/index.tsx:51]: Why is this all commented out?

File: web-client/src/shared/components/drivers-search/Filters/index.tsx [hunk 5]
```diff
@@ -1,7 +1,284 @@
...
+
+  const { loading, hasErrors, drivers } = useSelector(selectDrivers);
+
+  useEffect(() => {
+    dispatch(fetchDrivers());
+  }, [dispatch]);
+
+  if (loading) {
+    console.log('loading');
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/Filters/index.tsx:60]: Not sure if this is WIP, but generally console.log are for development purposes only (we try not include them in the final code unless we're dealing with code that is specifically meant for developers).

Thinking about this again, I think that the loading state might be more appropriate for the driver-search List component @WendyChenj . So maybe we don't need the loading state here.

File: web-client/src/pages/drivers-search.tsx (hunk 1 above)
Comment (OksanaSam) [reply web-client/src/pages/drivers-search.tsx:12]: Sorry, I haven't noticed that, I copied it over from my test file and didn't change back the title. Gonna change it back to what it was.

File: web-client/src/shared/components/drivers-search/Filters/filterData.js (hunk 2 above)
Comment (OksanaSam) [reply web-client/src/shared/components/drivers-search/Filters/filterData.js:1]: Sure, noted!

File: web-client/src/shared/components/drivers-search/Filters/index.tsx (hunk 5 above)
Comment (OksanaSam) [reply web-client/src/shared/components/drivers-search/Filters/index.tsx:60]: Yep, it's still WIP, I was checking if the redux state is updating properly. And yes, this state (all three, loading, hasErrors and drivers) will not be in this component at all, it will be in @WendyChenj component

File: web-client/src/pages/drivers-search.tsx (hunk 1 above)
Comment (OksanaSam) [reply web-client/src/pages/drivers-search.tsx]: Done, thanks

File: web-client/src/shared/components/drivers-search/Filters/index.tsx [hunk 6]
```diff
@@ -1,7 +1,284 @@
...
+  );
+  const [packingSortingFilter, setPackingSortingFilter] = useState(
+    packingSortingFilterdata,
+  );
+  const [insurancePolicyFilter, setInsurancePolicyFilter] = useState(
+    insurancePolicyFilterdata,
+  );
+  const [willingToLiftFilter, setWillingToLiftFilter] = useState(
+    willingToLiftFilterdata,
+  );
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/Filters/index.tsx:97]: Ah, I feel really bad about this, but I think this logic needs to be moved to the redux store.

It's a usability problem, and I should've realized this beforehand. Basically, I think we need the filters to be persistent. If a user leaves the search page for whatever reason, when ever they go back they should see their old filters. Using react's state management system, it's just not meant for persistence. Redux, on the other hand, has a lot of plugins that allow for more persistent structure.

What I would recommend is setting up the redux store like so

```
initState for driversSearch key in redux store
{
driversSearch: {
    filters: {
      [key]: DefaultValues (the ones you defined about)
    },
    response: {
      errors: Error[],
      data: Driver[],
      loading: Boolean
    }
  }
}
```
We won't worry about persistence just yet, because we'll be able to add that in later. For the reducer, I would add one really simple reducerFunction that can handle any kind of filter update, and three different reducer functions for the response key (one for loading, one for errors, and one for data)

File: web-client/src/shared/components/drivers-search/Filters/index.tsx [hunk 7]
```diff
@@ -1,7 +1,284 @@
...
+    //     //     console.log(Object.keys(item));
+    //     // }
+    // });
+    // OKSANA TO DO: to provide filterData as an argument to make a POST request
+    // dispatch(fetchDrivers("here"));
+    dispatch(fetchDrivers());
+  };
+
+  return (
+    // OKSANA TO DO: refactor this to map over one global state with all filters
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/Filters/index.tsx:188]: I like this idea a lot!

File: web-client/src/shared/components/drivers-search/Layout/index.tsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import React, { ReactNode } from 'react';
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/Layout/index.tsx:1]: Do you think you could create a branch on your machine (call it feature/Layout) and store this component there. Not sure we need it just yet.

File: web-client/src/shared/components/drivers-search/Layout/index.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+}: Props) => {
+    return (
+        <div>
+            <Head>
+                <title>{title}</title>
+                <meta charSet="utf-8" />
+                <meta name="viewport" content="initial-scale=1.0, width=device-width" />
+            </Head>
+            <Container>
+                <header>
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/Layout/index.tsx:23]: Perfect idiomatic use of tags here!

File: web-client/src/shared/redux/driversSearch/drivers.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  licenceClass: string;
+  vehicleType: string;
+  policeCheck: string;
+  drivingAbstract: string;
+  willingToLift: string;
+  packingAndSorting: string;
+  riskComfortLevel: string[];
+}
+
+export interface DriversState {
```
Comment (l2silver) [inline web-client/src/shared/redux/driversSearch/drivers.ts:16]: This would be the new response state in driverSearch key of redux store

File: web-client/src/shared/redux/driversSearch/drivers.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    },
+    getDriversSuccess: (state, { payload }) => {
+      state.drivers = payload;
+      state.loading = false;
+      state.hasErrors = false;
+    },
+    getDriversFailure: (state) => {
+      state.loading = false;
+      state.hasErrors = true;
+    }
```
Comment (l2silver) [inline web-client/src/shared/redux/driversSearch/drivers.ts:59]: These look good. I might change hasErrors to just errors, and instead of expecting a boolean, it would expect an array of Error objects. But other than that, these reducers are exactly what I had in mind. They should also point to the nested state.response instead of just state

File: web-client/src/shared/components/drivers-search/Filters/index.tsx (hunk 6 above)
Comment (OksanaSam) [reply web-client/src/shared/components/drivers-search/Filters/index.tsx:97]: Should it be a separate slice maybe? As they are not related at all

File: web-client/src/shared/components/drivers-search/Layout/index.tsx (hunk 8 above)
Comment (OksanaSam) [reply web-client/src/shared/components/drivers-search/Layout/index.tsx:1]: Ok, creating. I will delete it from here for now.

File: web-client/src/shared/redux/driversSearch/drivers.ts (hunk 10 above)
Comment (OksanaSam) [reply web-client/src/shared/redux/driversSearch/drivers.ts:16]: Btw, should licence class be an array too? -  licenceClass: string[]
And should we also add language here? language: string[]
They can speak a few languages, right?

File: web-client/src/shared/redux/driversSearch/drivers.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  }
+});
+
+// Actions generated from the drivers slice
+export const { getDrivers, getDriversSuccess, getDriversFailure } = driversSlice.actions;
+
+// Export state for using in useSelector 
+export const selectDrivers = (state: any) => state.driversSlice; // eslint-disable-line
+
+export const fetchDrivers = createAsyncThunk(
```
Comment (l2silver) [inline web-client/src/shared/redux/driversSearch/drivers.ts:69]: If I'm being totally honest, I'm amazed that you were able to figure out this createAsyncThunk. That being said, I hate it. I get what it does, but I think it just adds a lot of complexity for very little payout.

Checkout https://github.com/reduxjs/redux-thunk#motivation for a simple explanation of what a thunk is.

File: web-client/src/shared/redux/driversSearch/drivers.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+  'drivers/fetchDrivers',
+  async (filterData, thunkAPI) => {
+    thunkAPI.dispatch(getDrivers());
+    try {
+      const response = await fetch('/drivers', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(filterData)
```
Comment (l2silver) [inline web-client/src/shared/redux/driversSearch/drivers.ts:79]: excellent

File: web-client/src/shared/redux/driversSearch/drivers.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
+import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
```
Comment (l2silver) [inline web-client/src/shared/redux/driversSearch/drivers.ts:1]: I'm a bit confused. Why is there a driversSearch/drivers.ts and a driverSearch/index.ts

They look kind of similar.

File: web-client/src/shared/redux/index.ts [hunk 15]
```diff
@@ -2,14 +2,17 @@ import { configureStore } from '@reduxjs/toolkit';
...
-import driversSearchSlice from './driversSearch';
+// import driversSearchSlice from './driversSearch';
+import driversSlice from './driversSearch';
 
 const reducer = combineReducers({
-  driversSearch: driversSearchSlice.reducer
-})
+  // driversSearch: driversSearchSlice.reducer,
+  //OKSANA: this slice is making an API call
+  driversSlice
```
Comment (l2silver) [inline web-client/src/shared/redux/index.ts:11]: This can't be right? I'm looking at the old code on the left side, which looks more appropriate. Can you explain?

File: web-client/src/shared/components/drivers-search/Filters/filterData.js (hunk 2 above)
Comment (OksanaSam) [reply web-client/src/shared/components/drivers-search/Filters/filterData.js]: Renamed filterData.js to defaultValues.ts  and typed them

File: web-client/src/shared/redux/driversSearch/drivers.ts (hunk 14 above)
Comment (OksanaSam) [reply web-client/src/shared/redux/driversSearch/drivers.ts:1]: Yep, that driversSearch/drivers.ts is the old one, will be deleted. Sorry

File: web-client/src/shared/redux/index.ts (hunk 15 above)
Comment (OksanaSam) [reply web-client/src/shared/redux/index.ts:11]: Again WIP:) To explain: 
So in order not to mix stuff up, I've created my own slice - driversSlice - which does all the logic, I will rename it to driversSearch (meaning driversSearch: driversSearch). That one is exported from the slice file as driversSlice.reducer, so no need for that here and we can make use of ES6 object property value shorthand

File: web-client/src/shared/redux/driversSearch/drivers.ts (hunk 14 above)
Comment (OksanaSam) [reply web-client/src/shared/redux/driversSearch/drivers.ts]: Deleting done.

File: web-client/src/shared/components/drivers-search/Layout/index.tsx (hunk 8 above)
Comment (OksanaSam) [reply web-client/src/shared/components/drivers-search/Layout/index.tsx]: Done, Layout is out

File: web-client/public/mocks/handlers.js [hunk 16]
```diff
@@ -1,6 +1,31 @@
 import { rest } from 'msw'
-let mockDrivers = [{ id: 13, name: 'John Driver' }, { id: 22, name: 'Jim Driver'}, { id: 55, name: 'James Driver' }]
 
+const mockDrivers = [
```
Comment (l2silver) [inline web-client/public/mocks/handlers.js:3]: very cool! I love the coordination!

File: web-client/src/shared/redux/driversSearch/drivers.ts (hunk 10 above)
Comment (l2silver) [reply web-client/src/shared/redux/driversSearch/drivers.ts]: That's a great question. To be safe, lets make both of those string[]

File: web-client/src/shared/components/drivers-search/Filters/defaultValues.ts [hunk 17]
```diff
@@ -0,0 +1,99 @@
+export interface DefaultValue {
+  [name: string]: boolean | undefined
+}
+
+
+export const languageFilterdata: DefaultValue[] = [
```
[the remark this answers is not in the document]
Comment (l2silver) [reply web-client/src/shared/components/drivers-search/Filters/defaultValues.ts:6]: Likewise for the rest of the filters. Better to keep things as simple as possible.

[the remark this answers is not in the document]
Comment (OksanaSam) [reply web-client/src/shared/components/drivers-search/Filters/defaultValues.ts]: Thanks, done! All refactored to objects. Will also make one big object out of them and store it in redux.

File: web-client/src/shared/components/drivers-search/FilterAccordion/index.tsx [hunk 18]
```diff
@@ -11,21 +11,20 @@ import FormLabel from '@material-ui/core/FormLabel';
 import FormControl from '@material-ui/core/FormControl';
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/FilterAccordion/index.tsx:11]: Perfect!

File: web-client/src/shared/components/drivers-search/Filters/index.tsx [hunk 19]
```diff
@@ -1,7 +1,280 @@
...
+  //         hasErrors: state.hasErrors,
+  //         drivers: state.drivers
+  //     }
+  // });
+
+  const { loading, hasErrors, drivers } = useSelector(selectDrivers);
+
+  useEffect(() => {
+    dispatch(fetchDrivers());
+  }, [dispatch]);
```
Comment (l2silver) [inline web-client/src/shared/components/drivers-search/Filters/index.tsx:57]: Just noticed this. What's going on here?

File: web-client/src/shared/components/drivers-search/Filters/index.tsx (hunk 6 above)
Comment (l2silver) [reply web-client/src/shared/components/drivers-search/Filters/index.tsx:97]: I think we can leave it in the same slice.

File: web-client/src/shared/redux/index.ts (hunk 15 above)
Comment (l2silver) [reply web-client/src/shared/redux/index.ts:11]: Oh cool!

File: web-client/src/shared/components/drivers-search/Filters/index.tsx (hunk 19 above)
Comment (OksanaSam) [reply web-client/src/shared/components/drivers-search/Filters/index.tsx]: It was there just for testing, it's gone now

Comment (OksanaSam) [reply web-client/src/shared/components/drivers-search/Filters/index.tsx]: @l2silver Hey Leigh, could you please check (whenever you've got time) if this is how you wanna see thunk in the redux store. Since we are using createSlice, AppThunk, ThunkAction, and PayloadAction had to come into play (not sure if there is another way with slice). So it's still not the original thunk, but at least it looks more similar to that one. Thanks!

**Review by OksanaSam — COMMENTED** [review]
Still work in progress, have to make the filter work in order to send a proper request to the backend and move all local state to redux. The structure is built, have to move the functionality.

File: web-client/src/pages/api/drivers/index.ts [hunk 20]
```diff
@@ -0,0 +1,34 @@
+import type { NextApiRequest, NextApiResponse } from 'next';
```
Comment (l2silver) [inline web-client/src/pages/api/drivers/index.ts:1]: Thanks for doing this. Very clean!

**Review by l2silver — APPROVED** [review]
Hallelujah!
